### ISSUE TITLE
Domains are resolving despite being blocklisted

### ISSUE THREAD

**Issue by GBT7** [celzero/rethink-app#840]
Now that I've updated the app to 0.5.4a, I'm seeing logs of various ad domains resolving despite being on my on-device blocklists. In fact, the logs indicate no domains at all are getting blocked. I just tried switching to RDNS+ and now it appears the domains are still resolving despite being on my RDNS+ blocklists as well. How could that be?

**Comment by ignoramous** [comment]
This is a limitation on Android where all DNS requests are sent by the OS to the VPN (masking theidentity of the apps). Without treating DNS rules as IP rules, there's no way to allow / deny domains per-app.

To avoid confusion, in the upcoming version, `v054b`, we plan to tag domains as *may be blocked*:
- https://github.com/celzero/rethink-app/issues/817

**Comment by opk12** [comment]
Is it an option to return a fake address and postpone domain resolution to connection time? i.e. The DNS picks an address from the [black hole](https://en.wikipedia.org/wiki/Black_hole_(networking)#Dead_addresses), [private](https://en.wikipedia.org/wiki/Private_network), [link-local](https://en.wikipedia.org/wiki/Private_network#Link-local_addresses) ranges. When the firewall is triggered against the fake address, Rethink remembers the wanted domain and applies the app rules before doing the real DNS.

**Comment by ignoramous** [comment]
> Is it an option to return a fake address and postpone domain resolution to connection time?

This is slightly complicated, but we could implement this.

> The DNS picks an address from the [black hole](https://en.wikipedia.org/wiki/Black_hole_(networking)#Dead_addresses), [private](https://en.wikipedia.org/wiki/Private_network), [link-local](https://en.wikipedia.org/wiki/Private_network#Link-local_addresses) ranges. When the firewall is triggered against the fake address

This is what Rethink does today :) https://github.com/celzero/firestack/blob/8206cae1f3788b875fdfb52ae845aead94ac0528/intra/dnsx/alg.go#L115

**Comment by opk12** [comment]
Maybe it's paranoia but I'm also interested in this issue, to prevent fingerprinting and leaking OS version and manufacturer info through phoning home.

**Comment by opk12** [comment]
> The DNS picks an address
I meant to write *return* the wrong address to the app (i.e. fennec won't see the real github.com address)

**Comment by hussainmohd-a** [comment]
> > Apps aren't making those connections. Well, at least not if Rethink is doing what it is supposed to do.
> 
> This is the case for "allowed" apps, the problem is that now all the apps which I set to "bypass universal" are connecting when before they wouldn't.

The issue has been resolved and the fix will be incorporated in the upcoming release, ref: https://github.com/celzero/rethink-app/issues/816

**Comment by ignoramous** [comment]
`v054c` ([F-Droid](https://f-droid.org/en/packages/com.celzero.bravedns/), [Play Store](https://play.google.com/store/apps/details?id=com.celzero.bravedns)) has the fix for this. The DNS Logs UI now shows `maybe blocked` with `yellow` coloured border if the domain may be blocked by the *Firewall* depending on user-defined *trust* (allow) / *block* (deny) rules.

Feel free to reopen the bug in case you see this issue has not been fixed for you. Thanks.

**Comment by ThreeDeeJay** [comment]
@ignoramous So there's no way to prevent DNSs from getting resolved despite being blocked on both the blocklist and user rule? not even with adb/root/xposed or something?
![image](https://github.com/celzero/rethink-app/assets/71472458/0707aff2-b313-4aff-bb9f-75110d14c38e)
Like, aren't these unnecessary and unwanted connection that might prevent the device from sleeping and increase battery usage, and doesn't this mean we're leaking the domains we've blocked to the DNS server? 🤔

**Comment by ignoramous** [comment]
> doesn't this mean we're leaking the domains we've blocked to the DNS server

The screenshot you've shared is of a query to fetch *HTTPS Service Binding* records, which may contain data for establishing secure TLS connections. *HTTPS Service Binding* queries are not blocked unless the answer records come back with IP addresses. Whether those answer records have IPs or not cannot be known without resolving the query first.

**Comment by ThreeDeeJay** [comment]
@ignoramous I've also noticed it for other types of query like IPv4/6. 
So is this like a local-only query of DNS requests/metadata that doesn't actually get sent to the DNS server?
Otherwise, I'm just really confused as to why hostnames that are explicitly blocked would warrant any sort of online activity other than a null response to begin with 🤔
If that's the case, is this a limitation/behavior of Android/system VPN? and can we do anything about it?

**Comment by ignoramous** [comment]
> I've also noticed it for other types of query like IPv4/6.

Those must have been blocked after uncloaking the domain name (`CNAME`s aka "aliases" are in the blocklist but the original domain queried for isn't). Uncloaking a domain name can only happen after we get back answers from the upstream resolver.

Another scenario where a domain is resolved despite being in a blocklist is, either:
1. At least ONE app has been set to *Bypass DNS & Firewall* from *Configure* -> *Apps*
2. At least ONE app is *trusting* (*allow*) the domain in question.

> So is this like a local-only query of DNS requests/metadata that doesn't actually get sent to the DNS server?

Sent.

>  If that's the case, is this a limitation/behavior of Android/system VPN? and can we do anything about it?

If it happens for any scenario other than the ones outlined above (cloaked domain / non IPv4/IPv6 query, trusted domain, app bypassing dns & firewall), then it is a bug in Rethink.
